substrate/primitives/metadata-ir/src/v15.rs

```diff
@@ -17,30 +17,43 @@
 
 //! Convert the IR to V15 metadata.
 
-use crate::OuterEnumsIR;
+use crate::{OuterEnumsIR, TrasactionExtensionsByVersion};
 
 use super::types::{
 	ExtrinsicMetadataIR, MetadataIR, PalletMetadataIR, RuntimeApiMetadataIR,
 	RuntimeApiMethodMetadataIR, RuntimeApiMethodParamMetadataIR, TransactionExtensionMetadataIR,
 };
 
+use codec::Encode;
 use frame_metadata::v15::{
-	CustomMetadata, ExtrinsicMetadata, OuterEnums, PalletMetadata, RuntimeApiMetadata,
-	RuntimeApiMethodMetadata, RuntimeApiMethodParamMetadata, RuntimeMetadataV15,
-	SignedExtensionMetadata,
+	CustomMetadata, CustomValueMetadata, ExtrinsicMetadata, OuterEnums, PalletMetadata,
+	RuntimeApiMetadata, RuntimeApiMethodMetadata, RuntimeApiMethodParamMetadata,
+	RuntimeMetadataV15, SignedExtensionMetadata,
 };
+use scale_info::meta_type;
 
 impl From<MetadataIR> for RuntimeMetadataV15 {
 	fn from(ir: MetadataIR) -> Self {
+		const TRANSACTION_EXTENSIONS_BY_VERSION: &str = "transaction_extensions_by_version";
+		let transaction_extensions_by_version = ir.extrinsic.transaction_extensions_by_version();
+
 		RuntimeMetadataV15::new(
 			ir.pallets.into_iter().map(Into::into).collect(),
 			ir.extrinsic.into(),
 			ir.ty,
 			ir.apis.into_iter().map(Into::into).collect(),
 			ir.outer_enums.into(),
-			// Substrate does not collect yet the custom metadata fields.
-			// This allows us to extend the V15 easily.
-			CustomMetadata { map: Default::default() },
+			CustomMetadata {
+				map: [(
+					TRANSACTION_EXTENSIONS_BY_VERSION.into(),
+					CustomValueMetadata {
+						ty: meta_type::<TrasactionExtensionsByVersion>(),
+						value: transaction_extensions_by_version.encode(),
+					},
```

Comment:
I don't think this is enough, this encode the transaction extension indices used for each version.
But the v15 metadata extensions doesn't contains all the extensions.

If we have 2 version. v15 metadata will contains the first version only.
so we need to add in custom metadata all the other transaction extension, or just all the transaction extensions same as v16.

Reply:
Indeed, this will no longer work when we include new transactions in the next version.
We collect the same transaction extensions (signed extensions), but that will not be the case in the future. I'll introduce the transaction extensions as well, thanks for pointing this out 🙏